### DISCUSSION TITLE
Guidance on how to display a game file structure stored in a tree and edit its contents

### DISCUSSION THREAD

**Discussion by iMrShadow** [AvaloniaUI/Avalonia#17466] [Q&A]
Hello,

I am building an app that reads, modifies and writes game files. Due to some circumstances, I am deserialising them into a tree, and I am not using Reflection.

Here is how my node currently looks like (simplified):
``` cs
class TreeNode
 {
     string memberName; // Name of the field
     string name; // Type of the field 

     TreeNode parent; 
     List<TreeNode> children; 

     byte[] data; // The raw data

     bool isBlocked; // If the data blocked or not.
     bool isStruct; // If it's a struct only (either all data is stored here, if we don't know the contents, or none)
}
```

I want to potentially achieve something similar to [this](https://github.com/RicoSuter/VisualJsonEditor) or the [property grid extension](https://github.com/bodong1987/Avalonia.PropertyGrid) with some caveats:
1. In the files there are special flag fields, usually stored as a single integer. I want to have special GUI controls depending on the type of the flag. Let's say I have a simple integer - I want to be able to manipulate it using a checkbox, a combobox and a spinner.
2. Enums - in the example JSON project all enums are stored in the JSON file. However, it would be useful to pass the enums declared in a C# file.
3. It's a two-way operation. I am aware of converters, but I can't figure out how to handle flags and enums. There's also the fact that I may want to represent 2D/3D/4D vectors in the future.
4. I want to maintain the order of the members. The Property Grid extension uses categories, which I want to avoid.
5. Be able to add, insert or remove entries of array types, no matter what type of objects they store.

I had a couple of ideas, but I am not sure how possible or efficient they are. 

1. Manually generating components using a factory. The base class is a TreeNode and we use inheritance for the other types we want to parse. But not only this does not follow the MVVM pattern, but then I have no idea how to handle arrays, especially with insertion/deleting entries.
2. Have an intermediate JSON-like format. After reading through the whole file, parse every type of information into a JSON file. The flags are split into more intrinsic type controls (or an object), while the enums store all of their entries. But I don't think this is a very efficient solution, especially with how big some files can be.
3. Use a TreeViewModel, but so far I haven't found any examples where you can mix item templates. 

What would be the best approach? Any help or advice is appreciated, thanks

**Comment by timunie** [comment]
Maybe TreeDataGrid is for you?

**Reply by iMrShadow** [reply]
First of all, thanks for answering! However, I am not exactly sure if I explained it well enough.

I found a more relevant screenshot of what I want to achieve. 

![107880350-32981980-6eef-11eb-9939-2fce324890b9](https://github.com/user-attachments/assets/cbe95718-872e-4a0f-9f4f-376202c4fb7d)

In this example, let's say a parent node struct is `Transform3D`, while `Position`, `Rotation`, and `Scale` are its children. They are represented as 3D vectors, which you can edit the data each one with 3 sliders.
The `Material` struct has different types of children, for e.g. `Shininess` is a float, while `Ambient` is a 3D vector.

In my case - I want to make DataTemplates which inherit a tree node and work with the information. If the child is an integer - generate a spinner (`IntegerTreeNode`), if the child is a float - generate a slider (`FloatTreeNode`). If the child is an enum - generate a combobox with the values (`EnumTreeNode`). If it's a struct - create an expander containing its children (`StructTreeNode`). If it's an array - potentially represent them like [here](https://github.com/user-attachments/assets/ca63c490-9702-464b-be95-d73a3323bd9c).

Edit: I checked TreeDataGrid, but I didn't find any examples. However, I did inspect the [property value editor](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Diagnostics/Diagnostics/Views/PropertyValueEditorView.cs) from the debug tools. It did look promising at first, but I am not sure how to pass the TreeNode and work with it and potentially have some additional parameters

**Reply by timunie** [reply]
> Edit: I checked TreeDataGrid 

In the source of TDG there are some samples. Worth to check. 
- https://docs.avaloniaui.net/docs/reference/controls/treedatagrid/
- https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid 

Other than that you may need to provide a custom Style for your TreeView, you can look up the original from source: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/TreeView.xaml

**Reply by thevortexcloud** [reply]
Here is the docs on using tree view itself.

https://docs.avaloniaui.net/docs/reference/controls/treeview-1
